src/xrpld/ledger/detail/View.cpp

```diff
@@ -1088,6 +1223,121 @@ trustDelete(
+{
+    if (asset.holds<Issue>())
+    {
+        auto const& issue = asset.get<Issue>();
+        // Every account can hold XRP.
+        if (issue.native())
+            return tesSUCCESS;
+
+        auto const& issuerId = issue.getIssuer();
+        auto const& currency = issue.currency;
+        if (isGlobalFrozen(view, issuerId))
+            return tecFROZEN;
+
+        auto const& srcId = issuerId;
+        auto const& dstId = accountID;
+        auto const high = srcId > dstId;
+        auto const index = keylet::line(srcId, dstId, currency);
+        auto const sle = view.peek(keylet::account(accountID));
+        return trustCreate(
+            view,
+            high,
+            srcId,
+            dstId,
+            index.key,
+            sle,
+            /*auth=*/false,
+            /*noRipple=*/true,
+            /*freeze=*/false,
+            /*deepFreeze*/ false,
+            /*balance=*/STAmount{Issue{currency, noAccount()}},
+            /*limit=*/STAmount{Issue{currency, dstId}},
+            /*qualityIn=*/0,
+            /*qualityOut=*/0,
+            journal);
+    }
+
+    if (asset.holds<MPTIssue>())
+    {
+        auto const& mptIssue = asset.get<MPTIssue>();
+        auto const& mptID = mptIssue.getMptID();
+        auto const mpt = view.peek(keylet::mptIssuance(mptID));
+        if (mpt->getFlags() & lsfMPTLocked)
+            return tecLOCKED;
+        return MPTokenAuthorize::authorize(
+            view,
+            journal,
+            {.priorBalance = priorBalance,
+             .mptIssuanceID = mptID,
+             .accountID = accountID});
+    }
+
+    // Should be unreachable.
+    return tecINTERNAL;
+}
+
+[[nodiscard]] TER
+removeEmptyHolding(
+    ApplyView& view,
+    AccountID const& accountID,
+    Asset const& asset,
+    beast::Journal journal)
+{
+    if (asset.holds<Issue>())
+    {
+        auto const& issue = asset.get<Issue>();
+        if (issue.native())
+        {
+            auto const sle = view.read(keylet::account(accountID));
+            if (!sle)
+                return tecINTERNAL;
+            auto const balance = sle->getFieldAmount(sfBalance);
+            if (balance.xrp() != 0)
+                return tecHAS_OBLIGATIONS;
+            return tesSUCCESS;
+        }
+
+        // `asset` is an IOU.
+        auto const line = view.peek(keylet::line(accountID, issue));
+        if (!line)
+            return tecOBJECT_NOT_FOUND;
+        if (line->at(sfBalance)->iou() != beast::zero)
+            return tecHAS_OBLIGATIONS;
+        return trustDelete(
+            view,
+            line,
+            line->at(sfLowLimit)->getIssuer(),
+            line->at(sfHighLimit)->getIssuer(),
+            journal);
+    }
+
+    if (asset.holds<MPTIssue>())
+    {
+        auto const& mptIssue = asset.get<MPTIssue>();
+        auto const& mptID = mptIssue.getMptID();
+        // `MPTokenAuthorize::authorize` asserts that the balance is 0.
+        return MPTokenAuthorize::authorize(
```

Comment:
This function checks requireAuth for MPT but I don't see you checking that for IOU?